File: mc/rcmd/src/main/java/calebxzhou/rdi/mc/rcmd/RcmdArgument.java
package calebxzhou.rdi.mc.rcmd;

public record RcmdArgument<T>(String name, RcmdArgumentType<T> type) {
    public RcmdArgument {
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("参数名不能为空");
        }
        if (type == null) {
            throw new IllegalArgumentException("参数类型不能为空");
        }
    }
}

File: mc/rcmd/src/main/kotlin/calebxzhou/rdi/mc/rcmd/RcmdArgument.kt
package calebxzhou.rdi.mc.rcmd

data class RcmdArgument<T>(val name: String, val type: RcmdArgumentType<T>) {
    init {
        require(name.isNotBlank()) { "参数名不能为空" }
    }
}

data class RcmdArgumentParseResult<T>(val value: T, val nextTokenIndex: Int)

interface RcmdArgumentType<T> {
    fun name(): String

    @Throws(RcmdParseException::class)
    fun parse(tokens: List<String>, tokenIndex: Int): RcmdArgumentParseResult<T>
}
